### PR TITLE
F-71A User quota tests

### PR DESCRIPTION
**Pull request by aditipande96**
Signed-off-by: Aditi Pande <aditi.pande@seagate.com>

TD: https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/980090952/QA+F-71A+Capacity+usage+User+Level?focusedCommentId=988349337#comment-988349337

### CODE REVIEW

File: libs/s3/s3_misc.py [hunk 1]
```diff
@@ -163,6 +163,31 @@ def delete_objects_bucket(bucket_name, access_key: str, secret_key: str, **kwarg
...
+                                 endpoint_url=endpoint,
+                                 aws_access_key_id=access_key,
+                                 aws_secret_access_key=secret_key,
+                                 region_name=region,
+                                 **kwargs)
+    LOGGER.debug("S3 boto resource created")
+    resp_obj = s3_resource.Object(bucket_name, obj_name)
+    response = resp_obj.delete()
+    LOGGER.debug(response)
+    LOGGER.info("Object Deleted Successfully")
```
Comment (kachhwahadivya) [inline libs/s3/s3_misc.py:190]: Can u perfrom list object operation and check if the object is deleted and return True/False

File: tests/csm/rest/test_capacity_quota.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+
+from commons import configmanager
+from commons import cortxlogging
+from commons.utils import assert_utils
+from libs.csm.csm_interface import csm_api_factory
+from libs.csm.rest.csm_rest_quota import GetSetQuota
+from libs.s3 import s3_misc, s3_test_lib
+
+# pylint: disable=too-many-instance-attributes
+class TestSystemCapacity():
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:39]: Duplicate class name. We need unique name for each test class. It should be added to csm_interface.py

File: tests/csm/rest/test_capacity_quota.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    @classmethod
+    def setup_class(cls):
+        """ This is method is for test suite set-up """
+        cls.log = logging.getLogger(__name__)
+        cls.log.info("Initializing test setups ......")
+        cls.csm_obj = csm_api_factory("rest")
+        cls.cryptogen = SystemRandom()
+        cls.log.info("Initiating Rest Client ...")
+        cls.csm_conf = configmanager.get_config_wrapper(fpath="config/csm/test_rest_capacity.yaml")
+        cls.quota_obj = GetSetQuota()
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:51]: Can u use cls.csm_obj for all call, As mentioned above u need to add ur class to interface. U dont this object.

File: tests/csm/rest/test_capacity_quota.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+        cls.created_iam_users = set()
+        cls.s3_obj = s3_test_lib.S3TestLib()
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:53]: Do u really need this S3_obj. I see that u are s3_misc

File: tests/csm/rest/test_capacity_quota.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+        Teardowm method for deleting s3 account created in setup.
+        """
+        self.log.info("Deleting bucket %s & associated objects", self.bucket)
+        assert s3_misc.delete_objects_bucket(
+            self.bucket, self.akey, self.skey), "Failed to delete bucket."
+        self.log.info("Deleting buckets %s & associated objects", self.buckets_created)
+        buckets_deleted = []
+        iam_deleted = []
+        for bucket in self.buckets_created:
+            resp = s3_misc.delete_objects_bucket(bucket[0], bucket[1], bucket[2])
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:93]: Why do we have 2 delete_object_bucket?

File: tests/csm/rest/test_capacity_quota.py [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+    @pytest.mark.tags('TEST-40632')
+    def test_40632(self):
+        """
+        Test set & get API for User level quota & capacity admin user - enable
+        """
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info("Step 1: Creating IAM user with login as manage user")
+        payload = self.csm_obj.iam_user_payload_rgw("random")
+        resp = self.csm_obj.create_iam_user_rgw(payload)
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:130]: Can we create IAM user in setup_method?

File: tests/csm/rest/test_capacity_quota.py [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+        assert_utils.assert_true(resp.status_code == HTTPStatus.CREATED,
+                                 "IAM user creation failed")
+        self.created_iam_users.add(resp.json()['tenant'] + "$" + payload["uid"])
+        resp1 = self.csm_obj.compare_iam_payload_response(resp, payload)
+        self.log.info("Printing response %s", resp1)
+        assert_utils.assert_true(resp1[0], resp1[1])
+        self.log.info("Step 2: Create bucket under above IAM user")
+        self.akey = resp.json()["access_key"]
+        self.skey = resp.json()["secret_key"]
+        self.bucket_name = "iam-user-bucket-" + str(int(time.time()))
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:141]: time_ns

File: tests/csm/rest/test_capacity_quota.py [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+    @pytest.mark.cluster_user_ops
+    @pytest.mark.parallel
+    @pytest.mark.tags('TEST-40632')
+    def test_40632(self):
+        """
+        Test set & get API for User level quota & capacity admin user - enable
+        """
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info("Step 1: Creating IAM user with login as manage user")
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:128]: I think this should be admin user

File: tests/csm/rest/test_capacity_quota.py [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+        assert_utils.assert_true(resp4[0], resp4[1])
+        self.log.info("Step 5: Perform Put operation for 1 object of max size")
+        resp = s3_misc.create_put_objects(self.obj_name, self.bucket_name,
+                                          self.akey, self.skey, object_size=max_size)
+        assert_utils.assert_true(resp, "Put object Failed")
+        self.log.info("Step 6: Perform Put operation of Random size and 1 object")
+        random_size = self.cryptogen.randrange(1, max_size)
+        resp = s3_misc.create_put_objects(self.obj_name, self.bucket_name,
+                                          self.akey, self.skey, object_size=random_size)
+        assert_utils.assert_true(resp, "Put object Failed")
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:164]: Shouldnt this be assert_false?

File: tests/csm/rest/test_capacity_quota.py [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+        self.log.info("Step 3: Perform PUT API to set user level quota with random values")
+        uid = resp1.json()['tenant'] + "$" + payload["uid"]
+        test_cfg = self.csm_conf["test_40632"]
+        max_size = test_cfg["max_size"]
+        max_objects = test_cfg["max_objects"]
+        resp3 = self.quota_obj.set_user_quota(uid, "user","true", max_size, max_objects)
+        assert_utils.assert_true(resp3[0],resp3[1])
+        self.log.info("Step 4: Perform GET API to get user level quota")
+        resp4 = self.quota_obj.get_user_quota(uid, "user")
+        assert_utils.assert_true(resp4[0], resp4[1])
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:155]: Where is the verification for the get quota response matches SET quota?
Can we move SET and GET quota with verification in new function call?
Function which verifies that whatever is SET is visible in GET along with status code

File: tests/csm/rest/test_capacity_quota.py [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+        self.log.info("Step 8: Perform Put operation of small size and N object")
+        small_size = math.floor(max_size/max_objects)
+        for _ in range(0, max_objects):
+            resp = s3_misc.create_put_objects(self.obj_name, self.bucket_name,
+                                              self.akey, self.skey, object_size=small_size)
+            assert_utils.assert_true(resp, "Put object Failed")
+        self.log.info("Step 9: Perform Put operation of Random size and 1 object")
+        resp = s3_misc.create_put_objects(self.obj_name, self.bucket_name,
+                                          self.akey, self.skey, object_size=random_size)
+        assert_utils.assert_true(resp, "Put object Failed")
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:177]: Can we Delete all objects here

File: tests/csm/rest/test_capacity_quota.py [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+        self.log.info("##### Test ended -  %s #####", test_case_name)
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:178]: Can we have 2 more function 
Verify_max_size(max_size, object_cnt=1) - > perform put operation with 1 object and then perform put of random size and check if that fails.
Verify max_object(max_size,object_cnt) - perform put operation of N object of max_size/obj_cnt and check if one more put fails.

File: tests/csm/rest/test_capacity_quota.py [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+        self.log.info("Step 3: Perform PUT API to set user level quota with random values")
+        uid = resp1.json()['tenant'] + "$" + payload["uid"]
+        test_cfg = self.csm_conf["test_40634"]
+        max_size = test_cfg["max_size"]
+        max_objects = test_cfg["max_objects"]
+        resp3 = self.quota_obj.set_user_quota(uid, "user", "true", max_size, max_objects)
+        assert_utils.assert_true(resp3[0], resp3[1])
+        test_cfg = self.csm_conf["test_40634"]
+        for _ in range(0, test_cfg["num_iterations"]):
+            self.log.info("Step 4: Perform GET API to get user level quota")
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:283]: Can u print the iteration count

File: tests/csm/rest/test_capacity_quota.py [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+        for _ in range(0, max_objects):
+            resp = s3_misc.create_put_objects(self.obj_name, self.bucket_name,
+                                              self.akey, self.skey, object_size=small_size)
+            assert_utils.assert_true(resp, "Put object Failed")
+        self.log.info("Step 9: Perform Put operation of Random size and 1 object")
+        resp = s3_misc.create_put_objects(self.obj_name, self.bucket_name,
+                                          self.akey, self.skey, object_size=random_size)
+        assert_utils.assert_true(resp, "Put object Failed")
+        self.log.info("##### Test ended -  %s #####", test_case_name)
+
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:179]: Above comments are applicable for all the further tests.

File: tests/csm/rest/test_capacity_quota.py (hunk 3 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: used csm_obj instead of new object for class.

File: tests/csm/rest/test_capacity_quota.py (hunk 7 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:181]: changed

File: tests/csm/rest/test_capacity_quota.py (hunk 8 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 5 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:100]: removed.

File: config/csm/csm_config.yaml [hunk 15]
```diff
@@ -87,6 +87,7 @@ Restcall_LC:
   s3_iam_user_endpoint : "/api/v2/s3/iam/users"
   s3_iam_keys_endpoint : "/api/v2/s3/iam/keys"
   s3_iam_caps_endpoint : "/api/v2/s3/iam/caps"
+  get_set_quota: "/admin/user?quota&uid={}&quota-type={}"
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply config/csm/csm_config.yaml]: changed endpoint

File: libs/csm/rest/csm_rest_quota.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+            header = kwargs["headers"]
+        else:
+            header = self.headers
+        endpoint = CSM_REST_CFG["get_set_quota"]
+        endpoint = endpoint.format(uid, quota_type)
+        set_quota_payload = {"enabled": enabled,
+               "max_size": max_size,
+               "max_objects" : max_objects}
+        response = self.restapi.rest_call("put", endpoint=endpoint,
+                                          data=json.dumps(set_quota_payload),
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply libs/csm/rest/csm_rest_quota.py]: done

File: libs/s3/s3_misc.py (hunk 1 above)
Comment (aditipande96) [reply libs/s3/s3_misc.py:190]: list_object included

File: libs/csm/rest/csm_rest_quota.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+        :param header: header for api authentication
+        :param quota_type: Can be set to user or bucket
+        :return: response
+        """
+        self.log.info("Get IAM user request....")
+        if "headers" in kwargs.keys():
+            header = kwargs["headers"]
+        else:
+            header = self.headers
+        endpoint = CSM_REST_CFG["get_set_quota"]
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply libs/csm/rest/csm_rest_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 6 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 9 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: yes

File: tests/csm/rest/test_capacity_quota.py (hunk 11 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: After this step, buckets and objects are being deleted in teardown? Do we need explicit delete call here?

File: tests/csm/rest/test_capacity_quota.py (hunk 14 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:206]: okay

File: tests/csm/rest/test_capacity_quota.py (hunk 13 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:282]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 12 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:189]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 10 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: created the functions

File: config/csm/csm_config.yaml [hunk 18]
```diff
@@ -87,6 +87,7 @@ Restcall_LC:
   s3_iam_user_endpoint : "/api/v2/s3/iam/users"
   s3_iam_keys_endpoint : "/api/v2/s3/iam/keys"
   s3_iam_caps_endpoint : "/api/v2/s3/iam/caps"
+  get_set_quota: "/api/v2/iam/quota/{uid}"
```
Comment (kachhwahadivya) [inline config/csm/csm_config.yaml:90]: Can you make {} instead of {uid} as we will be forced to use UID variable

Comment (aditipande96) [reply config/csm/csm_config.yaml]: done.

**Review by kachhwahadivya — APPROVED** [review]

**Comment by RajeshPratapSInghChouhan** [comment]
please run and add logs of "pytest --collect-only --local True"

**Comment by aditipande96** [comment]
ERROR tests/security/test_cortx_port_scanner_kubectl_svc.py
ERROR tests/security/test_cortx_port_scanner_netstat.py
=================== 2127 tests collected, 2 errors in 11.68s ==================

**Review by RajeshPratapSInghChouhan — APPROVED** [review]

**Comment by aditipande96** [comment]
======= 2136 tests collected in 5.89s ==========

**Review by ketanarlulkar — APPROVED** [review]

**Review by ravindrachoudhari — APPROVED** [review]

**Pull request merged**